src/solver/mod.rs

```diff
@@ -35,17 +35,57 @@ struct AddClauseOutput {
 }
 
 /// Describes the problem that is to be solved by the solver.
-#[derive(Default)]
-pub struct Problem {
-    /// The requirements that _must_ have one candidate solvable be included in the
+///
+/// This struct follows the builder pattern and can have its fields set by one of the available
+/// setter methods.
```

Comment:
Maybe just add in comments why S is generic for future readers. Then it looks good to me.